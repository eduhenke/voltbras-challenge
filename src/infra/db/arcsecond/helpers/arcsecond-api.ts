import { RESTDataSource } from 'apollo-datasource-rest'

export default class ArcsecondApi extends RESTDataSource {
  constructor () {
    super()
    this.baseURL = 'https://api.arcsecond.io/'
  }

  async getRequest (uri: string): Promise<any> {
    const response = await this.get(`${this.baseURL}${uri}`)
    return JSON.parse(response)
  }
}
