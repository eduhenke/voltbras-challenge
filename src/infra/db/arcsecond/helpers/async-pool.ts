export async function asyncPool (poolLimit: number, array: number[], iteratorFn: any): Promise<any> {
  const ret = []
  const executing = []
  for (const item of array) {
    const p = Promise.resolve().then(() => iteratorFn(item, array))
    ret.push(p)

    if (poolLimit <= array.length) {
      const e = p.then(() => executing.splice(executing.indexOf(e), 1))
      executing.push(e)
      if (executing.length >= poolLimit) {
        await Promise.race(executing)
      }
    }
  }
  return Promise.all(ret)
}
