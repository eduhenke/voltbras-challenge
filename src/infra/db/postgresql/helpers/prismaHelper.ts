import { PrismaClient } from '@prisma/client'

export const PrismaHelper = {
  client: null as PrismaClient,

  async connect (): Promise<void> {
    this.client = new PrismaClient()
  },

  async disconnect (): Promise<void> {
    await this.client.$disconnect()
  }

}
