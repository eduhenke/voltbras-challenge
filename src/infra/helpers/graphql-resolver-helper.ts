export const graphqlResolverHelper = (useCase: any) => {
  return (_source: any, args: any, { dataSources }) => useCase(args,dataSources)
}
