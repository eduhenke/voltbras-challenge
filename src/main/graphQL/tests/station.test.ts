import {
  createTestClient,
  ApolloServerTestClient
} from 'apollo-server-testing'
import ArcsecondApi from '../../../infra/db/arcsecond/helpers/arcsecond-api'
import { typeDefs, resolvers } from '../'
import { ApolloServer } from 'apollo-server'
import { PrismaHelper } from '../../../infra/db/postgresql/helpers/prismaHelper'

beforeAll(async () => {
  await PrismaHelper.connect()
})

afterAll(async () => {
  await PrismaHelper.disconnect()
})

interface SutTypes {
  sut: ApolloServerTestClient
  arcsecondApiStub: ArcsecondApi
}

const makeSut = (): SutTypes => {
  const arcsecondApiStub = new ArcsecondApi()
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => {
      return {
        arcsecondApi: arcsecondApiStub
      }
    }
  })
  const sut = createTestClient(server)
  return {
    sut,
    arcsecondApiStub
  }
}

describe('Station', () => {
  test('Should add a new Station', async () => {
    const { sut } = makeSut()
    const INSTALL_STATION = `
      mutation InstallStation($input: StationInput!) {
        installStation(input: $input) {
          id,
          name,
          planet
        }
      }
    `
    const res = await sut.mutate({
      mutation: INSTALL_STATION,
      variables: { input: { name: 'any_name', planet: 'any_planet' } }
    })
    expect(res).toMatchSnapshot()
    expect(res.data.installStation.id).toBe('1')
    expect(res.data.installStation.name).toBe('any_name')
    expect(res.data.installStation.planet).toBe('any_planet')
  })
})
