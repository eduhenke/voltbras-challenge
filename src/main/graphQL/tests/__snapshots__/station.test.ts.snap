// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['Station Should add a new Station 1'] = `
Object {
  "data": Object {
    "installStation": Object {
      "id": "1",
      "name": "any_name",
      "planet": "2M1510A b",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`
